### PR TITLE
Add merge support per variable using cast style

### PR DESCRIPTION
**Pull request by rochacbruno**
When you have existing variables of type dict or list and you want to contribute to existing data instead of overriding
it is possible to use the special `@merge` notation.

Existing `settings.toml`
```toml
[default]
colors = ['yellow', 'blue', 'red']
```

Now you want to add `green` color to that existing list via environment variable:

```bash
export DYNACONF_COLORS="@merge ['green']"
```

When you access the variable you will get:

```py
from dynaconf import settings
assert settings.COLORS == ['green', 'yellow', 'blue', 'red']
```

But, what if green was already on the list? Dynaconf default behavior is to duplicate it like:

```toml
[default]
colors = ['yellow', 'blue', 'red', 'green']
```
```bash
export DYNACONF_COLORS="@merge ['green']"
```
```py
assert settings.COLORS == ['green', 'yellow', 'blue', 'red', 'green']
```

See the double `green`? how to avoid that? use **@merge_unique** which will not allow duplicated items, it will remove the duplicate from the existing list and add the new one in the new defined order.

```bash
export DYNACONF_COLORS="@merge_unique ['green']"
```
```py
assert settings.COLORS == ['yellow', 'blue', 'red', 'green']
```

The **@merge** also works for dictionaries:

```toml
[default]
database = {host="server.com", user="admin"}
```
```bash
export DYNACONF_DATABASE='@merge {password="s3cr4t"}'
```
```py
assert settings.DATABASE == {'host': 'server.com', 'user': 'admin', 'password': 's3cr4t'}
```

> NOTE: for dictionaries `@merge_unique` does not have any effect as it will by default exclude existing duplicate keys and replace with the newer ones.

#### Known caveats

The merge functionality works only for the **first** level keys on a dictionary and only **first** level items on a list.

In other words:

if I do `export DYNACONF_COLORS='@merge ["blue", "red", "green"]'`

```yaml
default:
  colors:
    - yellow
```

the exported variable will work and will merge to `colors` because it is the first level list on the `default` env and `settings.COLORS` is now `["blue", "red", "green", "yellow"]`.

However if I do if I do `export DYNACONF_COLORS='@merge {basic=["blue", "red", "green"]}'`

```yaml
default:
  colors:
    basic:
      - yellow
    compound:
      - fucsia
```

Now it is not merging the nested elements and you loose existing `yellow` value having `settings.BASIC.colors` as `["blue", "red", "green"]` and you kept `settings.COLORS.compound` unchanged.

### CODE REVIEW

**Comment by rochacbruno** [comment]
Hi @psav / @rsnyman I am planning to introduce this new thing and I would like to get your comments on it as you both worked on the `merge` feature.

I am adding this because of the need to solve an existing Django issue #111 and this way we can keep `MERGE_ENABLED_FOR_DYNACONF` disabled and merge only specific **first** level keys.

What do you guys think of this solution? any concern? any performance test you can do?

**Comment by rochacbruno** [comment]
@psav @rsnyman 

Future plan is to add something more clever for merge, without using the `@merge` notation, for future the idea is something like:

having 

```toml
[default]
database = {host="server.com", user="admin"}
```
I can do:

```bash
export DYNACONF_DATABASE='{dynaconf_merge=true, password=1234}'
```

And the presence of `dynaconf_merge=true` will mark that dictionary as a candidate for merging. (the same way we currently do with `dynaconf_include` token.

The same for lists `export DYNACONF_COLORS='["dynaconf_merge", "green"]'` and when `dynaconf_merge` item is present, the list is candidate for merging.

I would appreciate your input on this other idea also.

**Comment by rsnyman** [comment]
@rochacbruno I'll take a look and let you know.

**Comment by rochacbruno** [comment]
@bmbouter @dkliban your input would be also very helpful as you are also pontential end-users of this feature

**Comment by bmbouter** [comment]
@rochacbruno the feature as described in [this comment](https://github.com/rochacbruno/dynaconf/pull/135#issuecomment-476880207) makes a lot of sense to me.

**Comment by rochacbruno** [comment]
@bmbouter @dkliban @psav @rsnyman 

Closing in favor of #139 (better implementation)
